### PR TITLE
[spirv] Support GPU subgroup MMA ops in SPIRVVectorizeLoadStore

### PR DESCRIPTION
**Pull request by antiagainst**
This change enables vectorizing memref types used by GPU subgroup MMA ops so that we can be coherent and allow casting vector<8xf16> values into vector<4xf32> throughout.

### CODE REVIEW

**Comment by antiagainst** [comment]
Depends on https://reviews.llvm.org/D137143

File: compiler/src/iree/compiler/Codegen/SPIRV/SPIRVVectorizeLoadStore.cpp [hunk 1]
```diff
@@ -419,6 +436,72 @@ class ProcessInterfaceBindingSubspan final
...
+    }
+
+    Location loc = storeOp.getLoc();
+    auto indices = adjustIndices(scalarMemrefType, vectorMemrefType,
+                                 adaptor.getIndices(), rewriter, loc);
+    if (failed(indices)) return failure();
+
+    auto newLeadDimSize = rewriter.getIntegerAttr(
+        storeOp.getLeadDimensionAttr().getType(),
+        vectorMemrefType.getDimSize(vectorMemrefType.getRank() - 1));
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/SPIRV/SPIRVVectorizeLoadStore.cpp:496]: I don't think we should rely on the stride of the memref, we should follow whatever is in the op and just convert based on the type change.

File: compiler/src/iree/compiler/Codegen/SPIRV/SPIRVVectorizeLoadStore.cpp [hunk 2]
```diff
@@ -73,6 +75,11 @@ static unsigned calculateMemRefVectorNumBits(
     SmallVectorImpl<Operation *> &uses) {
   unsigned minBits = kMaxVectorNumBits;
   for (Operation *op : uses) {
+    if (isa<gpu::SubgroupMmaLoadMatrixOp, gpu::SubgroupMmaStoreMatrixOp>(op)) {
+      // GPU subgroup MMA ops do not care about the memref element type. It can
+      // load from whatever memref element type.
+      continue;
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/SPIRV/SPIRVVectorizeLoadStore.cpp:81]: one case we would not be able to change is if the stride is not divisible by the new memref type size. I believe it is possible so would be good to check for it to be sure we don't miscompile.

File: compiler/src/iree/compiler/Codegen/SPIRV/SPIRVVectorizeLoadStore.cpp (hunk 1 above)
Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/SPIRV/SPIRVVectorizeLoadStore.cpp]: Good points! That makes sense. This actually caught a bug. Thanks!

File: compiler/src/iree/compiler/Codegen/SPIRV/SPIRVVectorizeLoadStore.cpp (hunk 2 above)
Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/SPIRV/SPIRVVectorizeLoadStore.cpp]: Yeah good catch! Sorry this was sort of done in a hurry..

**Review by ThomasRaoux — APPROVED** [review]

**Pull request merged**